File: src/C#/BehavioralPatterns/State/VendingMachine/DispenseProductState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace VendingMachine
{
    public class DispenseProductState : State
    {
        public DispenseProductState(VendingMachine vendingMachine) : base(vendingMachine)
        {
            Console.WriteLine("DISPENSING");
        }

        public override void InsertMoney(decimal amount)
            => Console.WriteLine("Cannot insert money during the dispensing of the product");

        public override void SelectProduct(string productCode) => Console.WriteLine("The product is already selected.");

        public override void DispenseProduct()
        {
            if (vendingMachine.SelectedProductCode == null)
            {
                Console.WriteLine("There is no selected product for dispensing");
                vendingMachine.SetState(new IdleState(vendingMachine));
                return;
            }

            Console.WriteLine("Dispensing product.");
            Thread.Sleep(2000);

            var product = vendingMachine.Products.FirstOrDefault(p => p.Code == vendingMachine.SelectedProductCode);
            product.Stock--;

            vendingMachine.SelectedProductCode = null;

            Console.WriteLine("The product is dispensed.");

            if (vendingMachine.Products.All(p => p.Stock == 0))
            {
                vendingMachine.SetState(new SoldOutState(vendingMachine));
            }
            else
            {
                vendingMachine.SetState(new IdleState(vendingMachine));
            }
        }

        public override void Cancel() => Console.WriteLine("Cannot cancel the dispensing operation");

        public override void Refill(List<Product> products) =>
            Console.WriteLine("Cannot refill during the dispensing of the product");
    }
}

File: src/JavaScript/BehavioralPatterns/State/VendingMachine/dispenseProductState.js

const IdleState = require("./idleState");
const State = require("./state");
const SoldOutState = require("./soldOutState");

class DispenseProductState extends State {
    constructor(vendingMachine) {
        super(vendingMachine);
        console.log("DISPENSING");
    }

    insertMoney(amount) {
        console.log("Cannot insert money during the dispensing of the product");
    }

    selectProduct(productCode) {
        console.log("The product is already selected.");
    }

    dispenseProduct() {
        if (!this.vendingMachine.selectedProductCode) {
            console.log("There is no selected product for dispensing");
            this.vendingMachine.setState(new IdleState(this.vendingMachine));
            return;
        }

        console.log("Dispensing product.");
        setTimeout(() => {
            const productIndex = this.vendingMachine.products.findIndex(p => p.code === this.vendingMachine.selectedProductCode);
            if (productIndex !== -1) {
                this.vendingMachine.products[productIndex].stock--;
                console.log("The product is dispensed.");
                if (this.vendingMachine.products.every(p => p.stock === 0)) {
                    this.vendingMachine.setState(new SoldOutState(this.vendingMachine));
                } else {
                    this.vendingMachine.setState(new IdleState(this.vendingMachine));
                }
            }
        }, 2000);
    }

    cancel() {
        console.log("Cannot cancel the dispensing operation");
    }

    refill(products) {
        console.log("Cannot refill during the dispensing of the product");
    }
}

module.exports = DispenseProductState;